### ISSUE TITLE
'recv' unexpectedly blocking leads to Processor hang crash

### ISSUE THREAD

**Issue by mtalexander** [SDL-Hercules-390/hyperion#377]
I'm testing MTS in SDL Hercules.  If I leave it up but inactive for a few hours the watchdog thread will eventually catch a hung CPU. I tracked this down to the following problem in the `console_connect` thread:

The function `console_connection_handler` gets the device lock for the 3270 being used for the MTS operator's console.  It then calls `recv_3270_data` to read something from the 3270.  There is a call to `recv` in it which doesn't return immediately, so the thread waits with the device lock held.

MTS later tries to start a channel program on the 3270.  This hangs the CPU since the start subchannel command can't get the device lock.

I'm not sure exactly what the problem is. There shouldn't be anything useful to read from the 3270 at this time since the system is idle and I was asleep. Should it maybe _not_ be calling `recv_3270_data`?  Should maybe `recv_3270_data` do something to see if any data is available and return immediately if not?  Should it maybe release the device lock before calling `recv_3270_data` and then get it again later? If so, what happens if someone else mucks with the device in the meantime?

This problem probably affects the MTS operator's console because MTS will periodically write something to it to update status information.  This happens completely asynchronously even if nothing else is going on at the console.  This is more unusual than for other uses of 3270 devices, including as an MTS terminal.  This is probably why I've only seen this on the operator's console device and not on a terminal device.

**Comment by Fish-Git** [comment]
Without even looking at the code you mention, I have to say I am very pleased that you took the time and effort to try and debug the problem for yourself! Thank you! We need more users like you. &nbsp; `:)`

I cannot answer any of your questions at this time however, since I have not reviewed any of the code you mentioned yet. Let me take a look at it and get back to you.

In the mean time, did you review the **"[Submitting problem reports](https://github.com/SDL-Hercules-390/hyperion/issues/373)**" issue (#373) before submitting this/your problem report? I'm guessing you didn't since you failed to include many of the items listed there.

Please take a look at it and then supply the missing information in a subsequent comment:

* Hercules version
* Hercules config
* Hercules log
* Host system (Windows? Linux? Mac? Other?)
* Etc...

Thanks!

**Comment by Fish-Git** [comment]
Okay, I just reviewed the logic in question and am not seeing what you're seeing.

The device lock is only held long enough to `recv` the data that we _know_ for a fact is available, meaning the `recv` should _not_ be blocking. The `recv` _should_ complete pretty much _immediately_ since we know for a fact that at least _some_ data is available. (Why else would `recv_3270_data` be called in the first place?!)

I really need to know what version of Hercules you are using, etc. (i.e. I need you to answer all the questions (provide all of the information) as described previously and as documented in **"[Submitting problem reports](https://github.com/SDL-Hercules-390/hyperion/issues/373)**".

Thanks.

**Comment by mtalexander** [comment]
Sorry about leaving out all that info.  I should know better.

Hercules version: 004e0b87 with changes to add MTS HIM support (none of which are anywhere near that code).  The full version string is "4.4.9999.0-SDL-gde7e7931 (4.4.9999.0)" but you won't find that SHA1 in your repo.

Config file and log file attached.

Host system Mac running Mohave

I've also attached the crash log which contains the stack trace for the console_connect thread that shows it is waiting in a call to recvfrom.

I see what you're saying, but something isn't quite right.  The device lock is obtained and released while building the `fdset` for the call to `pselect` around line 3522 in console.c.  It is then obtained again around line 3700 and not released until around line 3775 after the call to `recv_3270_data` .  Somewhere between the call to `pselect` and the call to `recv_3270_data` something is happening that means there is nothing to be read.  `recv_3270_data` is clearly blocking with the device lock set, the question is why.  

There were two or three other 3270 sessions connected to Hercules at this time.  None of these should have had any data to be sent, but perhaps the process of dealing with them caused problems with the 3270 used for the operator.  Maybe there needs to be a `pselect` call inside the device loop that starts around line 3690 just to be sure there is still data ready.  That would probably fix the problem, whatever it is.  FWIW I'm using the Zoc8 emulator for TN3270 support but I doubt that matters.
      
[Archive.zip](https://github.com/SDL-Hercules-390/hyperion/files/6134292/Archive.zip)

**Comment by Fish-Git** [comment]
> I've also attached the crash log which contains the stack trace for the console_connect thread that shows it is waiting in a call to _**`recvfrom`**_.

_Well, **there's** your problem!!_ &nbsp; `:-o`

Why is _`recvfrom`_ being called?!

It _should_ be calling _`recv`!_ &nbsp; (_not_ `recvfrom`!)

Methinks this is a likely Apple Mac bug. &nbsp; (Or possibly a Zoc8 bug?)

> I'm using the Zoc8 emulator for TN3270 support but I doubt that matters.

I'm not so sure....

Looking at their [**changes log**](http://www.emtec.com/downloads/zoc/zoc_changes.txt), it looks like they _did_ have a problem they had to fix that may have been related to [**`recvfrom`**](https://pubs.opengroup.org/onlinepubs/009696699/functions/recvfrom.html) _(which **is** completely different from plain [**`recv`**](https://pubs.opengroup.org/onlinepubs/009695399/functions/recv.html) after all)_:

> <pre>
> VERSION 8.01.6 (2020-12-17)
> FIX: telnet connections did sometimes block main window when reverse-dns lookup took too long
> </pre>

But IMHO it is more than likely an Apple Mac problem somewhere. It should _NOT_ be calling `recvfrom`. It should be calling _**`recv`**_.

**Bottom line:** this does _not_ appear to be a Hercules problem.

p.s. &nbsp;Are you _this_ Mike Alexander? &nbsp; `:)`

* https://www.eecis.udel.edu/~mills/gallery/gallery8.html

**Comment by mtalexander** [comment]
That backtrace entry is not `recvfrom`, it's `_recvfrom`.  External symbols on the Mac have a single "_" prepended.  This symbol starts with "__" (two underscores), hence the original symbol was `_recvfrom`.

Also, `recv` and `recvfrom` aren't all that different.  From the man page that covers both of them

     The recv() function is normally used only on a connected socket (see
     connect(2) or connectx(2)) and is identical to recvfrom() with a null
     pointer passed as its address argument.  As it is redundant, it may not
     be supported in future releases.

Both block if no data is available.  From later in the man page:

    If no messages are available at the socket, the receive call waits for a
     message to arrive, unless the socket is nonblocking (see fcntl(2)) in
     which case the value -1 is returned and the external variable errno set
     to EAGAIN.

Furthermore, recv_3270_data is actually calling recv.  Here's the relevant code:

    53ca: 48 8d 85 e0 ff fe ff         	leaq	-65568(%rbp), %rax
    53d1: b9 00 00 00 00               	movl	$0, %ecx
    53d6: ba 00 00 01 00               	movl	$65536, %edx
    53db: 48 89 c6                     	movq	%rax, %rsi
    53de: e8 00 00 00 00               	callq	0 <_recv_3270_data+0xca>
    53e3: 89 45 ec                     	movl	%eax, -20(%rbp)
    53e6: 83 7d ec 00                  	cmpl	$0, -20(%rbp)
    53ea: 0f 89 45 01 00 00            	jns	325 <_recv_3270_data+0x21c>

The relevant relocation entry is 

   00000000000053df X86_64_RELOC_BRANCH _recv

My guess is that both `recv` and `recvfrom` call `_recvfrom` to do the real work.   The first, `recv`, passes a null pointer for socket address and `recvfrom` passes the address it got.  This is done is such a way that somebody (the compiler or linker) optimized out the intermediate call and just sent the caller directly to `_recvfrom` (with the underscore prefix).

I still say that this is not a Mac bug.  There is some circumstance in which the result from the (much) earlier call to pselect is not still valid when the test around line 3728 is made.

It's also not a Zoc problem.  Zoc is at the far end of a TCP connection.  There is no way it could cause a call to `recv` to become a call to `recvfrom` in the host it's talking to.  That change log entry seems to be about a hang at the client end, not a hang on the machine Zoc is talking to.

Yes, that picture is me (about 50 years ago).  I've changed.  :-)

**Comment by Fish-Git** [comment]
> That backtrace entry is not `recvfrom`, it's `_recvfrom`. External symbols on the Mac have a single "_" prepended. This symbol starts with "__" (two underscores), hence the original symbol was `_recvfrom`.

Knew that. &nbsp; `:)`

> Also, `recv` and `recvfrom` aren't all that different. From the man page that covers both of them
> 
> > _The recv() function is normally used only on a connected socket (see connect(2) or connectx(2)) and is identical to recvfrom() with a null pointer passed as its address argument.  As it is redundant, it may not be supported in future releases._

Fair enough.

> It's also not a Zoc problem. Zoc is at the far end of a TCP connection. There is no way it could cause a call to `recv` to become a call to `recvfrom` in the host it's talking to. That change log entry seems to be about a hang at the client end, not a hang on the machine Zoc is talking to.

Fair enough again.

> I still say that this is not a Mac bug.

And I still say it _is!_ &nbsp; `>;-)`

> There is some circumstance in which the result from the (much) earlier call to `pselect` is not still valid when the test around line 3728 is made.

Which to _me_ clearly indicates a Mac bug!

(Or something else unusual going on with your system somewhere! Maybe some special network code code somewhere? Antivirus or something perhaps, that is maybe wanting to examine the data being read from the socket before it allows it to be delivered to the caller? Maybe some code somewhere in the Mac doing something weird/unexpected like that maybe?)

The return from the (_not_ "much" earlier!) `pselect` said there was data waiting to be read, but then when we go to read it, it's not there! For some unknown reason, Apple's (or _someone's!_) code has decided to discard the data that _was_ there when the call to `pselect` was done.

Network (socket) data should not just _disappear_, Mike!

And Hercules is clearly not the one responsible for making the data disappear either IMO. The `recv_3270_data` function is the _only_ place in `console.c` 3270 logic where data is being read from the terminal (once it's been connected that is), and it is only ever called whenever `pselect` indicates there is data waiting to be read!

So you tell me Mike, how in the world this can possibly be a Hercules problem?! Because I'm just not seeing it!

Based on all of the evidence you've provided, this is _clearly_ (to me!) a Mac bug! For some unknown reason, under certain unknown circumstances, your Mac (or possibly your new MTS HIM code?) or some other code somewhere, is causing the data that _was_ there, to suddenly disappear!

**Comment by Fish-Git** [comment]
> The return from the (_not_ "much" earlier!) `pselect` said there was data waiting to be read, but then when we go to read it, it's not there! For some unknown reason, Apple's (or _someone's!_) code has decided to discard the data that _was_ there when the call to `pselect` was done.

Either that or there wasn't any data waiting to be read in the first place. That is to say it _could_ be a bug in `pselect` that's causing it to indicate data is available for a given socket when in reality there actually isn't any. That's a possibility too: a bug in `pselect`.

But my point is, this does _**NOT**_ appear to be a Hercules bug/issue.

Rather, this very much appears to be an _Apple Mac_ bug/issue.

**Comment by mtalexander** [comment]
This conversation seems to be trending toward a metaphysical discussion of "what is a bug".  I think we agree that `recv` is blocking when `pselect` says it shouldn't.  I suspect we'll never know why and at some level it doesn't matter.  

I found at least one description of a very similar problem on  [Stack Overflow](https://stackoverflow.com/questions/27667672/no-data-available-to-recv-after-select).  It's a bit different (UDP, not TCP and called from Python) but it sure sounds like the same thing.  Even if we determine what the bug is we probably can't fix it.  It's probably in Apple code or perhaps Zoc, neither of which we have any control over.

I think we should change Hercules to cope with this and not crash.  I say this for two reasons: First, programs like Hercules (and MTS) should not crash when external errors occur.  I've spent lots of time working on error handling over the years and I think it has mostly been time well spent.

Second, it's arguably wrong to call a routine that might block, even if it shouldn't block in this case,  when holding a lock unless that's unavoidable.  It's not unavoidable in this case.  I suggest that Hercules should make the 3270 socket non-blocking (either around the recv call or generally) and deal with the error return if no data is available.  I'll make this change and submit a pull request.  I'll probably make the socket non-blocking temporarily around the `recv` call since that avoids having to change all `recv` calls and I don't think the overhead will be prohibitive.

**Comment by Fish-Git** [comment]
After reading your reply, I did some more digging and found this:

* https://man7.org/linux/man-pages/man2/select.2.html#BUGS

> On Linux, select() may report a socket file descriptor as "ready
> for reading", while nevertheless a subsequent read blocks.  This
> could for example happen when data has arrived but upon
> examination has the wrong checksum and is discarded.  There may
> be other circumstances in which a file descriptor is spuriously
> reported as ready.  Thus it may be safer to use O_NONBLOCK on
> sockets that should not block.

Incredible! &nbsp;`:-(`

So I'm going to have to agree with you on this one Mike: given that the bug _is_ known to exist, and _is_ ultimately outside of Hercules's control (and is unlikely to be addressed/fixed any time soon given that it's been known about for many _years_ now! (Sheesh!)), and _does_ cause serious, _catastrophic_ problems whenever it does happen to strike, yes, I agree: Hercules definitely _should_ be enhanced to be ... less susceptible to? ... this specific _known_ operating system bug.

I apologize for my giving you a hard time, but I have a natural reluctance to introduce "workarounds" (band-aid fixes) for bugs in other peoples' code. I'd much rather pressure them instead to fix their damn bug.

Anyway, I look forward to eventually seeing your proposed workaround.

**Comment by Fish-Git** [comment]
p.s. Given that the description of the `pselect` bug mentions it is caused by the received data having a bad checksum, have you checked to see if the problem still occurs when you use a different tn3270 client other than Zoc? Maybe the problem _is_ with Zoc? Or maybe tuntap?

I'm just trying to understand why, if this bug has been known to exist for so many years now, it hasn't bitten anyone else until now. Why only you? Why just Apple Mac? Why don't other Linux users using other *Nix distributions apparently not have this problem?

It might be worth a try to try a different 3270 client other than Zoc and/or to check for any fixes to tuntap, etc.

_&lt;shrug&gt;_

Just a thought. &nbsp;`:)`

**Comment by Fish-Git** [comment]
> (and is unlikely to be addressed/fixed any time soon given that it's been known about for many _years_ now! (Sheesh!))

Apparently, this bug has existed in Linux _since at least **2012!**_

* https://web.archive.org/web/20120314211713/https://man7.org/linux/man-pages/man2/select.2.html#BUGS

**Comment by wrljet** [comment]
> 
> I'm just trying to understand why, if this bug has been known to exist for so many years now, it hasn't bitten anyone else until now. Why only you? Why just Apple Mac? Why don't other Linux users using other *Nix distributions apparently not have this problem?
> 

Not to put too fine a point on it, but Apple isn't Linux.  They are Unix derived.

**Comment by Fish-Git** [comment]
> Not to put too fine a point on it, but Apple isn't Linux. They are Unix derived.

Unix, Linux, whatever... It's all "*nix" to me,&nbsp; `:)`

My point is, if this bug _is/was_ in *nix (and evidence suggests that it is/was), it's apparently one that most _Linux_ distributions have either since fixed (given that none our _Linux_ users have ever reported being bitten by it before (and we have _many more_ Linux users than we do Apple Unix users)), or else there's something unique to Mike's system that's triggering it more frequently for him. That is to say, _regardless_ of where the bug is, why is it apparently being tripped so consistently and frequently for Mike? (whereas it's not for any of our other *nix users?)

The Hercules code that is affected by it has been quite stable and un-problematic for a _very_ long time, so if the bug _did_ exist unfixed in Linux and/or Unix, you would think that more than one of our Linux users would have reported it by now. But they haven't. Thus it makes sense to try and help Mike find out why it's apparently biting him and why now? What changed? What's he doing differently from virtually all of our other *nix users? What's different?

Thus my suggestion for Mike to try a different 3270 client and/or research whether there is an update to tuntap for his system (or maybe a buggy network adapter???). There's obviously _something_ that's causing these checksum errors (presumably) to occur more frequently on his system than on anyone else's, so it makes sense to me that we should try attacking the problem from _that_ angle.  _&lt;shrug&gt;_

Yes, sure, I fully agree that Hercules should be "fix" (hardened) so that when such an error does occur, Hercules doesn't crash as a result of it. That I agree 100% with Mike about. But why is _HE_ experiencing this problem? That's what I'm trying to concentrate on right now. (I presume he's working on the Hercules workaround so he can continue with his MTS HIM development). Installing the workaround into Hercules is only going to prevent the crash from occurring whenever the problem does occur. It's only going to _mask_ the problem, but it's _**not**_ going to fix the _**root**_ of the problem: why it's occurring in the first damn place!

I prefix _fixing_ problems, not hiding from them.

**Comment by mtalexander** [comment]
I agree with all of that.  There are a few things I can add that might help narrow down the problem.

First, this is happening on the loopback interface (Zoc and Hercules are in the same machine connected via localhost).  I don't think tun/tap is involved, in fact I don't think it's possible for an ordinary mortal to configure a tun/tapp interface on a Mac, at least I haven't found out how.  Apple has locked down the kernel to the point that things like this are extremely difficult to do.  The tunnelblick folks have a page about this [here](https://www.tunnelblick.net/cTunTapConnections.html) .

I also don't think it's a checksum problem, although that's just a guess.  The man page says there are other cases in which this can happen and a checksum error on the loopback interface seems unlikely.  The checksum is calculated by the same code that is checking it in the localhost network stack.  I doubt that Zoc is involved in calculating the checksum.

I think the thing that is different about my environment is MTS.  No one else is running that version of MTS right now, and few are running any version.  So far every occurrence has happened when I've had at least 3 tn3270 connections open to Hercules and it's always happened on the operator's console.  I'm sure that the software that drives the operator's console does some things with a 3270 which are unusual.  

Furthermore it has so far only happened when there are two MTS tasks using the same 3270.  Mts has a device command to start a second session on the same 3270.  Only one of them can be active at the same time, and I don't see how this could affect the interaction between the host and the 3270, but this correlation is interesting.

I coded up the fix last night and it's running now.  So far it hasn't logged any failure.  I want to see at least a few failures before I declare this code working.

By the way I want to thank you for adding such complete logging to the watchdog thread.  When this first happened I figured I would have to run Hercules under the debugger to find the cause, but it was pretty easy to find from the information logged.

**Comment by Fish-Git** [comment]
> Mts has a device command to start a second session on the same 3270. Only one of them can be active at the same time, and I don't see how this could affect the interaction between the host and the 3270, but this correlation is interesting.

Quite!

Have you tried enabling CCW (I/O) tracing on all of your MTS terminals yet? If not, you should.

I'm presuming the following are your MTS terminals?
```
621-62B     UDPH     10.0.0.46
62C-643     TLNT     10.0.0.46
644-64F     TCPH     10.0.0.46
```

Thus, before IPLing MTS (or, at the minimum, just before you do whatever it is that triggers the problem), you would then do:
```
t+621
t+622
... etc...
t+64e
t+64f
```
Since there are so many you might even make it a simple .rc script to make it easier to do, and then just enter the command `script traceon.rc` to enable tracing or `script traceoff.rc` to disable (where the traceoff script would of course do `t-xxx` for all of the same terminals).

You might also consider enabling the Console.c debugging build option if you haven't already done so:
```diff
--- hyperion-1/console.c	2021-03-12 21:47:11.585351400 -0800
+++ hyperion-0/console.c	2021-03-16 18:07:01.518831400 -0700
@@ -64,7 +64,7 @@
 #define DEBUG_PROGRESS_TOO  2       /* debug progress too            */
 #define DEBUG_EVERYTHING    3       /* debug buffers too             */
 
-#define DEBUG_LVL     DEBUG_NOTHING /* Choose your poison */
+#define DEBUG_LVL     DEBUG_EVERYTHING /* Choose your poison */
 
 #if DEBUG_LVL == DEBUG_NOTHING
   #define CONERROR                          WRMSG
```
That should produce a bunch of debugging messages for each terminal I/O, one or more of which might hopefully help shed some additional light on matter.

_&lt;shrug&gt;_

Again, just suggestions. &nbsp;`:)`

**Comment by Fish-Git** [comment]
> Mts has a device command to start a second session on the same 3270. Only one of them can be active at the same time, and I don't see how this could affect the interaction between the host and the 3270, but this correlation is interesting.

Another thought: even though you say "Only one of them can be active at the same time", maybe the problem is just an unfortunate timing issue? Maybe the `pselect` says "this terminal has some data to be read" but before it can be read, some _other_ MTS thread/task issues a _write_ to the same terminal that causes whatever data that _was_ there to be discarded? (due to the write?)

I forget all of the many 3270 commands that a 3270 device accepts/understands (I used to know but I've since forgotten), but as I recall whenever a field is modified, the field attribute is automatically updated to indicate "this field has been modified" and the data can then be read via a "read modified" command to read _just_ that field (instead of the entire freaking screen).

But what if some _write_ I/O sneaks in ahead of the `recv` to cause Zoc to toss everything away and reset everything? (such that a "read modified" would then not have anything to return?)

Could maybe something like be occurring?

That's why I think doing a CCW trace and/or enabling the console.c debugging build option might be helpful.

And once again, these are just suggestions! You do your debugging in whatever way works best for you!

**Comment by mtalexander** [comment]
I created pull request 379 with a fix for this.  It took longer because I was waiting for nother failure to make sure it fixed the problem and, as luck would have it, I didn't get a failure for a while.  

I tried to make as small a change as possible.  The socket is only made non-blocking for the one recv call that is causing problems.  It would be possible to make it non-blocking all the time but this is a much bigger change and more likely to introduce further bugs.  It doesn't seem worth the effort or risk.

**Comment by Fish-Git** [comment]
Mike: (@mtalexander)

Please see Jürgen and my comments regarding your Pull Request. I need to know whether you want to make my suggested change or, as Jürgen suggested, leave it in. Either way is fine by me! But as your pull will be committed with your name, I need you to let me know before I do it.

Thanks!!

**Comment by Fish-Git** [comment]
"Fixed" (workaround implemented) by Pull Request #379. Closing.
